plugins/inputs/firehose/firehose_request.go

```diff
@@ -0,0 +1,140 @@
+package firehose
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal"
+)
+
+type record struct {
+	EncodedData string `json:"data"`
+}
+
+type requestBody struct {
+	RequestID string   `json:"requestId"`
+	Timestamp int64    `json:"timestamp"`
+	Records   []record `json:"records"`
+}
+
+// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#requestformat
+type request struct {
+	req  *http.Request
+	body requestBody
+	res  response
+}
+
+// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#responseformat
+type response struct {
+	body       responseBody
+	statusCode int
+}
+
+type responseBody struct {
+	RequestID    string `json:"requestId"`
+	Timestamp    int64  `json:"timestamp"`
+	ErrorMessage string `json:"errorMessage,omitempty"`
+}
+
+func (r *request) authenticate(expected config.Secret) error {
+	if expected.Empty() {
+		return nil
+	}
+	key := r.req.Header.Get("x-amz-firehose-access-key")
+	match, err := expected.EqualTo([]byte(key))
+	if err != nil {
+		return fmt.Errorf("comparing keys failed: %w", err)
```

Comment:
You don't set a status code here! What happens with the response to the sender? Shouldn't this be a 500? Maybe you should make sure that in `sendResponse` (or when creating the request) you will set a defined fallback status code (maybe 500?)!

Reply:
Oh I missed that. And agreed with the fallback status code